# Importing Libraries

In [2]:
# Import libraries
import pandas as pd
import numpy as np
import os

# Importing dataframes

In [3]:
# Creating a path variable
path = r'C:\Users\Aaron\CareerFoundry\Instacart Basket Analysis'

In [4]:
# Importing orders_products_merged
ords_prods_merge = pd.read_pickle(os.path.join(path, '02 Data', 'Prepared Data', 'orders_products_merged_UPDATED.pkl'))

# Creating a subset of the first one million entries

In [5]:
df = ords_prods_merge[:1000000]

In [6]:
# Checking the new subset
df.shape

(1000000, 18)

In [7]:
# Checking the new subset
df.head()

order_id  user_id  order_number  orders_day_of_week  order_time  \
0  2539329.0      1.0           1.0                 2.0         8.0   
1  2398795.0      1.0           2.0                 3.0         7.0   
2   473747.0      1.0           3.0                 3.0        12.0   
3  2254736.0      1.0           4.0                 4.0         7.0   
4   431534.0      1.0           5.0                 4.0        15.0   

   days_since_prior_order  product_id  add_to_cart_order  reordered _merge  \
0                    11.0         196                1.0        0.0   both   
1                    15.0         196                1.0        1.0   both   
2                    21.0         196                1.0        1.0   both   
3                    29.0         196                1.0        1.0   both   
4                    28.0         196                1.0        1.0   both   

  product_name  aisle_id  department_id  prices    price_range_loc  \
0         Soda      77.0            7.0     9.0  Mid-range product   
1         Soda      77.0            7.0     9.0  Mid-range product   
2         Soda      77.0            7.0     9.0  Mid-range product   
3         Soda      77.0            7.0     9.0  Mid-range product   
4         Soda      77.0            7.0     9.0  Mid-range product   

      busiest_day    busiest_days busiest_period_of_day  
0  Regularly busy  Regularly busy        Average orders  
1  Regularly busy    Slowest days        Average orders  
2  Regularly busy    Slowest days           Most orders  
3      Least busy    Slowest days        Average orders  
4      Least busy    Slowest days           Most orders

# Splitting the data using groupby()

In [8]:
df.groupby('product_name')

# Aggregating data with agg()

In [9]:
# Performing a single aggregation
df.groupby('department_id').agg({'order_number': ['mean']})

order_number
                      mean
department_id             
4.0              18.825780
7.0              17.472355
13.0             17.993423
14.0             19.246334
16.0             19.463012
17.0             11.294069
19.0             19.305237
20.0             17.599636

In [10]:
# Performing multiple aggregation
df.groupby('department_id').agg({'order_number': ['mean', 'min', 'max']})

order_number           
                      mean  min   max
department_id                        
4.0              18.825780  1.0  99.0
7.0              17.472355  1.0  99.0
13.0             17.993423  1.0  99.0
14.0             19.246334  1.0  99.0
16.0             19.463012  1.0  99.0
17.0             11.294069  1.0  98.0
19.0             19.305237  1.0  99.0
20.0             17.599636  1.0  99.0

# Aggregating data with transform()

In [11]:
ords_prods_merge['max_order'] = ords_prods_merge.groupby(['user_id'])['order_number'].transform(np.max)

In [12]:
# Checking the results
ords_prods_merge.head(15)

order_id  user_id  order_number  orders_day_of_week  order_time  \
0   2539329.0      1.0           1.0                 2.0         8.0   
1   2398795.0      1.0           2.0                 3.0         7.0   
2    473747.0      1.0           3.0                 3.0        12.0   
3   2254736.0      1.0           4.0                 4.0         7.0   
4    431534.0      1.0           5.0                 4.0        15.0   
5   3367565.0      1.0           6.0                 2.0         7.0   
6    550135.0      1.0           7.0                 1.0         9.0   
7   3108588.0      1.0           8.0                 1.0        14.0   
8   2295261.0      1.0           9.0                 1.0        16.0   
9   2550362.0      1.0          10.0                 4.0         8.0   
10  2968173.0     15.0          15.0                 1.0         9.0   
11  1870022.0     15.0          17.0                 2.0        16.0   
12  1911383.0     15.0          18.0                 2.0        11.0   
13  2715276.0     15.0          21.0                 1.0         9.0   
14   487368.0     15.0          22.0                 1.0        10.0   

    days_since_prior_order  product_id  add_to_cart_order  reordered _merge  \
0                     11.0         196                1.0        0.0   both   
1                     15.0         196                1.0        1.0   both   
2                     21.0         196                1.0        1.0   both   
3                     29.0         196                1.0        1.0   both   
4                     28.0         196                1.0        1.0   both   
5                     19.0         196                1.0        1.0   both   
6                     20.0         196                1.0        1.0   both   
7                     14.0         196                2.0        1.0   both   
8                      0.0         196                4.0        1.0   both   
9                     30.0         196                1.0        1.0   both   
10                     7.0         196                2.0        0.0   both   
11                     8.0         196                6.0        1.0   both   
12                     7.0         196                1.0        1.0   both   
13                     7.0         196                1.0        1.0   both   
14                    14.0         196                1.0        1.0   both   

   product_name  aisle_id  department_id  prices    price_range_loc  \
0          Soda      77.0            7.0     9.0  Mid-range product   
1          Soda      77.0            7.0     9.0  Mid-range product   
2          Soda      77.0            7.0     9.0  Mid-range product   
3          Soda      77.0            7.0     9.0  Mid-range product   
4          Soda      77.0            7.0     9.0  Mid-range product   
5          Soda      77.0            7.0     9.0  Mid-range product   
6          Soda      77.0            7.0     9.0  Mid-range product   
7          Soda      77.0            7.0     9.0  Mid-range product   
8          Soda      77.0            7.0     9.0  Mid-range product   
9          Soda      77.0            7.0     9.0  Mid-range product   
10         Soda      77.0            7.0     9.0  Mid-range product   
11         Soda      77.0            7.0     9.0  Mid-range product   
12         Soda      77.0            7.0     9.0  Mid-range product   
13         Soda      77.0            7.0     9.0  Mid-range product   
14         Soda      77.0            7.0     9.0  Mid-range product   

       busiest_day    busiest_days busiest_period_of_day  max_order  
0   Regularly busy  Regularly busy        Average orders       10.0  
1   Regularly busy    Slowest days        Average orders       10.0  
2   Regularly busy    Slowest days           Most orders       10.0  
3       Least busy    Slowest days        Average orders       10.0  
4       Least busy    Slowest days           Most orders       10.0  
5   Regularly busy  Regularly 

# Deriving columns with loc()

In [13]:
ords_prods_merge.loc[ords_prods_merge['max_order'] > 40, 'loyalty_flag'] = 'Loyal Customer'

In [14]:
ords_prods_merge.loc[(ords_prods_merge['max_order'] <= 40) & (ords_prods_merge['max_order'] > 10), 'loyalty_flag'] = 'Regular customer'

In [15]:
ords_prods_merge.loc[ords_prods_merge['max_order'] <= 10, 'loyalty_flag'] = 'New customer'

In [16]:
ords_prods_merge['loyalty_flag'].value_counts(dropna = False)

Regular customer    15891507
Loyal Customer      10294027
New customer         6249525
NaN                       11
Name: loyalty_flag, dtype: int64

In [17]:
# Showing the first 30 of specific columns, onlyl user_id, max_order, and loyalty_flag
ords_prods_merge[['user_id', 'max_order', 'loyalty_flag']].head(30)

user_id  max_order      loyalty_flag
0       1.0       10.0      New customer
1       1.0       10.0      New customer
2       1.0       10.0      New customer
3       1.0       10.0      New customer
4       1.0       10.0      New customer
5       1.0       10.0      New customer
6       1.0       10.0      New customer
7       1.0       10.0      New customer
8       1.0       10.0      New customer
9       1.0       10.0      New customer
10     15.0       22.0  Regular customer
11     15.0       22.0  Regular customer
12     15.0       22.0  Regular customer
13     15.0       22.0  Regular customer
14     15.0       22.0  Regular customer
15     19.0        9.0      New customer
16     19.0        9.0      New customer
17     19.0        9.0      New customer
18     21.0       33.0  Regular customer
19     31.0       20.0  Regular customer
20     31.0       20.0  Regular customer
21     43.0       11.0  Regular customer
22     43.0       11.0  Regular customer
23     52.0       27.0  Regular customer
24     52.0       27.0  Regular customer
25     52.0       27.0  Regular customer
26     52.0       27.0  Regular customer
27     52.0       27.0  Regular customer
28     52.0       27.0  Regular customer
29     52.0       27.0  Regular customer

# Inserting Exercise 4.9 Cleaning steps here:

In [18]:
# Let's find out if there are any values above a certain threshold, anything above "normal" supermarket price
ords_prods_merge.loc[ords_prods_merge['prices'] > 100]

order_id   user_id  order_number  orders_day_of_week  order_time  \
10030345   912404.0      17.0          12.0                 2.0        14.0   
10030346   603376.0      17.0          22.0                 6.0        16.0   
10030347  3264360.0     135.0           2.0                 2.0        21.0   
10030348   892534.0     135.0           3.0                 0.0         8.0   
10030349   229704.0     342.0           8.0                 1.0        19.0   
...             ...       ...           ...                 ...         ...   
29194168  2249946.0  204099.0          29.0                 0.0         8.0   
29194169  2363282.0  204099.0          31.0                 0.0         9.0   
29194170  3181945.0  204395.0          13.0                 3.0        15.0   
29194171  2486215.0  205227.0           7.0                 3.0        20.0   
29194172  2232716.0  205227.0          10.0                 0.0        11.0   

          days_since_prior_order  product_id  add_to_cart_order  reordered  \
10030345                     5.0       21553                5.0        0.0   
10030346                     4.0       21553                3.0        1.0   
10030347                    13.0       21553                6.0        0.0   
10030348                    12.0       21553                3.0        1.0   
10030349                    30.0       21553                9.0        0.0   
...                          ...         ...                ...        ...   
29194168                     4.0       33664                1.0        0.0   
29194169                     2.0       33664                1.0        1.0   
29194170                     8.0       33664               25.0        0.0   
29194171                     4.0       33664                8.0        0.0   
29194172                     4.0       33664               23.0        1.0   

         _merge                      product_name  aisle_id  department_id  \
10030345   both  Lowfat 2% Milkfat Cottage Cheese     108.0           16.0   
10030346   both  Lowfat 2% Milkfat Cottage Cheese     108.0           16.0   
10030347   both  Lowfat 2% Milkfat Cottage Cheese     108.0           16.0   
10030348   both  Lowfat 2% Milkfat Cottage Cheese     108.0           16.0   
10030349   both  Lowfat 2% Milkfat Cottage Cheese     108.0           16.0   
...         ...                               ...       ...            ...   
29194168   both             2 % Reduced Fat  Milk      84.0           16.0   
29194169   both             2 % Reduced Fat  Milk      84.0           16.0   
29194170   both             2 % Reduced Fat  Milk      84.0           16.0   
29194171   both             2 % Reduced Fat  Milk      84.0           16.0   
29194172   both             2 % Reduced Fat  Milk      84.0           16.0   

           prices     price_range_loc     busiest_day    busiest_days  \
10030345  14900.0  High-range product  Regularly busy  Regularly busy   
10030346  14900.0  High-range product  Regularly busy  Regularly busy   
10030347  14900.0  High-range product  Regularly busy  Regularly busy   
10030348  14900.0  High-range product     Busiest day    Busiest days   
10030349  14900.0  High-range product  Regularly busy    Busiest days   
...           ...                 ...             ...             ...   
29194168  99999.0  High-range product     Busiest day    Busiest days   
29194169  99999.0  High-range product     Busiest day    Busiest days   
29194170  99999.0  High-range product  Regularly busy    Slowest days   
29194171  99999.0  High-range product  Regularly busy    Slowest days   
29194172  99999.0  High-range product     Busiest day    Busiest days   

         busiest_period_of_day  max_order      loyalty_flag  
10030345           Most orders       40.0  Regular customer  
10030346           Most orders       40.0  Regular customer  
10030347        Average orders        4.0      New customer  
10030348        Average orders        4.0      New customer  
1

We can see that there are 5,127 items that are greater than 100

Additionally, we can see the 2 prices greater than 100 are: 14,900 and 99,999

##### For now, let's mark these high prices as missing since they don't make sense in terms of how high they are compared to the rest of the values in the prices column.

In [19]:
ords_prods_merge.loc[ords_prods_merge['prices'] > 100, 'prices'] = np.nan

###### Now we check the max price in the prices column to see what it is now that we've replaced the 14,900 and 99,999 values

In [21]:
ords_prods_merge['prices'].max()

25.0

Logically, a max value of 25 dollars in a grocery store makes more sense than a value of 14,900 or 99,999 dollars

# Exporting updated dataframes

In [24]:
ords_prods_merge.to_pickle(os.path.join(path, '02 Data', 'Prepared Data', 'ords_prods_merge_NEW.pkl'))